### PR TITLE
Delete instance api

### PR DESCRIPTION
**Pull request by antho1404**
dependency: https://github.com/mesg-foundation/engine/pull/1036

### CODE REVIEW

**Comment by ilgooz** [comment]
service:delete deletes the service even if an instance is running which is the opposite behavior in the proposal

File: protobuf/api/instance.proto [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  rpc Delete (DeleteInstanceRequest) returns (DeleteInstanceResponse) {}
+}
+
+message CreateInstanceRequest {
+  string serviceHash = 1;      // Service's sid or hash.
+  repeated string env = 2;     // Env vars to apply to service's instance on runtime.
+}
+
+message CreateInstanceResponse {
+  string hash = 2;        // Service's instance hash.
```
Comment (krhubert) [inline protobuf/api/instance.proto]: why starts from 2?

File: protobuf/api/instance.proto [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  string serviceHash = 3; // Service's bare hash.
+}
+
+message DeleteInstanceRequest {
+  string hash = 1;      // Instance hash
+}
+
+message DeleteInstanceResponse {
+  string hash = 1;      // Instance hash
```
Comment (krhubert) [inline protobuf/api/instance.proto]: why return instance hash in response if it sent in the request

File: database/instance_db.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// Close closes database.
+func (d *LevelDBInstanceDB) Close() error {
+	return d.db.Close()
+}
+
+// Delete deletes service from database.
+func (d *LevelDBInstanceDB) Delete(hash string) error {
+	tx, err := d.db.OpenTransaction()
```
Comment (krhubert) [inline database/instance_db.go]: the key can be deleted with a single [delete](https://godoc.org/github.com/syndtr/goleveldb/leveldb#DB.Delete) call

File: database/instance_db.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	// encode service
+	b, err := d.marshal(i)
+	if err != nil {
+		tx.Discard()
+		return err
+	}
+
+	// save instance with hash.
+	if err := tx.Put([]byte(i.Hash), b, nil); err != nil {
```
Comment (krhubert) [inline database/instance_db.go]: the key can be saved with a single call without using transaction

File: database/instance_db.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	return &s, nil
+}
+
+// Get retrives instance by instance hash.
+func (d *LevelDBInstanceDB) Get(hash string) (*instance.Instance, error) {
+	tx, err := d.db.OpenTransaction()
+	if err != nil {
+		return nil, err
+	}
+	b, err := tx.Get([]byte(hash), nil)
```
Comment (krhubert) [inline database/instance_db.go]: the key can be get with a single call without using transaction

File: instance/instance.go [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package instance
+
+// Instance is an instance of a service.
+// This contains a reference to a service that is running.
+// Multiple instances can run for the same service as long as they have different configurations
+type Instance struct {
+	Hash        string `hash:"-"`
+	ServiceHash string `hash:"serviceHash"`
```
Comment (krhubert) [inline instance/instance.go]: instance tags are not needed here (structhash is not used on the struct)

File: sdk/instance/instance.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+type Instance struct {
+	container  container.Container
+	serviceDB  database.ServiceDB
+	instanceDB database.InstanceDB
+}
+
+// New creates a new Instance SDK with given options.
+func New(c container.Container, serviceDB database.ServiceDB, instanceDB database.InstanceDB) *Instance {
+	return &Instance{
+		container:  c,
```
Comment (krhubert) [inline sdk/instance/instance.go]: sdk should use the manager as a bridge to container package

File: database/instance_db.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
+package database
```
Comment (krhubert) [inline database/instance_db.go]: There are no tests for this package

File: protobuf/api/instance.proto (hunk 1 above)
Comment (antho1404) [reply protobuf/api/instance.proto]: moved to the right PR

File: database/instance_db.go (hunk 4 above)
Comment (antho1404) [reply database/instance_db.go]: move to the right pr

File: database/instance_db.go (hunk 3 above)
Comment (antho1404) [reply database/instance_db.go]: fixed

File: protobuf/api/instance.proto (hunk 2 above)
Comment (antho1404) [reply protobuf/api/instance.proto]: I don't mind much about that, I can remove but the idea is that when we delete a resource we have the id of the resource delete in order to be able to have a full asynchrony application that can on one side send the request and on the other side just receive the result and know how to update a local store or anything.

But again I'm happy to delete it if needed

**Pull request merged**